feat: Document custom arguments (#553)
PaperMC/docs: docs/paper/dev/api/command-api/basics/custom-arguments.mdx

        }

        ctx.getSource().getSender().sendRichMessage("Player <player> is an operator!",
            Placeholder.component("player", player.displayName())
        );
        return Command.SINGLE_SUCCESS;
    })
```

As you can see, there is a ton of logic not directly involved with the functionality of the command. And if we want to use this same argument on another node, we have to
copy-paste a lot of code. It goes without saying that this would be incredibly tedious.

The solution to this problem are custom arguments. Before going into detail about them, this is how the argument would look when implemented as a custom argument:

```java title="OppedPlayerArgument.java"
@NullMarked
public final class OppedPlayerArgument implements CustomArgumentType<Player, PlayerSelectorArgumentResolver> {

    @Override
    public Player parse(StringReader reader) {
        throw new UnsupportedOperationException("This method will never be called.");
    }

    @Override
    public <S> Player parse(StringReader reader, S source) throws CommandSyntaxException {
        if (!(source instanceof CommandSourceStack stack)) {

>             final Message message = MessageComponentSerializer.message().serialize(text("The source needs to be a CommandSourceStack!"));

Reviewer: ```suggestion
            final Message message = MessageComponentSerializer.message().serialize(Component.text("The source needs to be a CommandSourceStack!"));
```

            throw new CommandSyntaxException(new SimpleCommandExceptionType(message), message);
        }

        final Player player = getNativeType().parse(reader).resolve(stack).getFirst();
        if (!player.isOp()) {

>             final Message message = MessageComponentSerializer.message().serialize(text(player.getName() + " is not a server operator!"));

Reviewer: ```suggestion
            final Message message = MessageComponentSerializer.message().serialize(Component.text(player.getName() + " is not a server operator!"));
```

    }

    @Override
    public ArgumentType<PlayerSelectorArgumentResolver> getNativeType() {
        return ArgumentTypes.player();
    }

    @Override
    public <S> CompletableFuture<Suggestions> listSuggestions(CommandContext<S> ctx, SuggestionsBuilder builder) {
        Bukkit.getOnlinePlayers().stream()
            .filter(ServerOperator::isOp)
            .map(Player::getName)
            .filter(name -> name.toLowerCase(Locale.ROOT).startsWith(builder.getRemainingLowerCase()))
            .forEach(builder::suggest);
        return builder.buildFuture();
    }
}
```

At a first look, that seems like way more code than it was needed to just do the logic in the command tree itself. So what is the advantage?
The answer becomes apparent rather quickly when we look at how the argument is now declared:

```java
Commands.argument("player", new OppedPlayerArgument())
    .executes(ctx -> {
        final Player player = ctx.getArgument("player", Player.class);

        ctx.getSource().getSender().sendRichMessage("Player <player> is an operator!",
            Placeholder.component("player", player.displayName())
        );
        return Command.SINGLE_SUCCESS;
    })
```

> This is way more readable and easy to understand than without custom argument. And it is reusable! Hopefully, you now have a basic grasp of **why** you should use custom arguments.

Reviewer: ```suggestion
This is way more readable and easy to understand when using a custom argument. And it is reusable! Hopefully, you now have a basic grasp of **why** you should use custom arguments.
```

## Examining the `CustomArgumentType` interface
The interface is declared as follows:

```java title="CustomArgumentType.java"
package io.papermc.paper.command.brigadier.argument;

@NullMarked
public interface CustomArgumentType<T, N> extends ArgumentType<T> {

    @Override
    T parse(final StringReader reader) throws CommandSyntaxException;

    @Override
    default <S> T parse(final StringReader reader, final S source) throws CommandSyntaxException {
        return ArgumentType.super.parse(reader, source);
    }

    ArgumentType<N> getNativeType();

    @Override
    @ApiStatus.NonExtendable
    default Collection<String> getExamples() {
        return this.getNativeType().getExamples();
    }

    @Override
    default <S> CompletableFuture<Suggestions> listSuggestions(final CommandContext<S> context, final SuggestionsBuilder builder) {
        return ArgumentType.super.listSuggestions(context, builder);
    }
}
```

### Generic types
There are three generic types present in the interface:

> - `T`: This is the type of the class that is returned when `CommandContext#getArgument` is called on this argument
> - `N`: The native type of the class which this custom argument extends. Used as the "underlying" argument

Reviewer: ```suggestion
- `T`: This is the type of the class that is returned when `CommandContext#getArgument` is called on this argument.
- `N`: The native type of the class which this custom argument extends. Used as the "underlying" argument.
```

- `S`: A generic type for the command source. Will usually be a `CommandSourceStack`.

### Methods
| Method declaration                                                                                                              | Description  |
|---------------------------------------------------------------------------------------------------------------------------------|--------------|

> | `ArgumentType<N> getNativeType()`                                                                                               |  Here, you declare the underlying argument type which is used as a base for clientside argument validation                                                                                             |

Reviewer: ```suggestion
| `ArgumentType<N> getNativeType()`                                                                                               |  Here, you declare the underlying argument type, which is used as a base for client-side argument validation.                                                                                             |
```

> | `T parse(final StringReader reader) throws CommandSyntaxException`                                                              |  This method is used if `T parse(StringReader, S)` was not overwritten. In here you can run conversion and validation logic.                                                                             |

Reviewer: ```suggestion
| `T parse(final StringReader reader) throws CommandSyntaxException`                                                              |  This method is used if `T parse(StringReader, S)` is not overridden. In here, you can run conversion and validation logic.                                                                             |
```

> | `default <S> T parse(final StringReader reader, final S source)`                                                                |  If overwritten, this method will be preferred to `T parse(StringReader)`. It serves the same purpose, but allows to also   include the source in the parsing logic.

Reviewer: ```suggestion
| `default <S> T parse(final StringReader reader, final S source)`                                                                |  If overridden, this method will be preferred to `T parse(StringReader)`. It serves the same purpose, but allows including the source in the parsing logic.
```

> | `default Collection<String> getExamples()`                                                                                      |  This method should **not** be extended. It is used internally to differentiate certain argument types while parsing.                                                                             |

Reviewer: ```suggestion
| `default Collection<String> getExamples()`                                                                                      |  This method should **not** be overridden. It is used internally to differentiate certain argument types while parsing.                                                                             |
```

> | `default <S> CompletableFuture<Suggestions> listSuggestions(final CommandContext<S> context, final SuggestionsBuilder builder)` |  This method is the equivalent of `RequiredArgumentBuilder#suggests(SuggestionProvider<S>)`. You can overwrite this method in order to send your own suggestions to the client. |

Reviewer: ```suggestion
| `default <S> CompletableFuture<Suggestions> listSuggestions(final CommandContext<S> context, final SuggestionsBuilder builder)` |  This method is the equivalent of `RequiredArgumentBuilder#suggests(SuggestionProvider<S>)`. You can override this method in order to send your own suggestions to the client. |
```

### A very basic implementation
```java
package io.papermc.commands;

import com.mojang.brigadier.StringReader;
import com.mojang.brigadier.arguments.ArgumentType;
import com.mojang.brigadier.arguments.StringArgumentType;
import io.papermc.paper.command.brigadier.argument.CustomArgumentType;
import org.jspecify.annotations.NullMarked;

@NullMarked
public class BasicImplementation implements CustomArgumentType<String, String> {

    @Override
    public String parse(StringReader reader) {
        return reader.readUnquotedString();
    }

    @Override
    public ArgumentType<String> getNativeType() {
        return StringArgumentType.word();
    }
}
```

Notice the use of `reader.readUnquotedString()`. In addition to allowing existing argument types to parse your argument,
you can also manually read input. Here, we read an unquoted string, the same as a word string argument type.

## `CustomArgumentType.Converted<T, N>`
In case that you need to parse the native type to your new type, you can instead use the `CustomArgumentType.Converted` interface.

> This interface is an extension to the `CustomArgumentType` interface, which adds two new, overwritable methods:

Reviewer: ```suggestion
This interface is an extension to the `CustomArgumentType` interface, which adds two new, overridable methods:
```

                builder.suggest(flavor.toString());
            }
        }

        return builder.buildFuture();
    }

    @Override
    public ArgumentType<String> getNativeType() {
        return StringArgumentType.word();
    }
}
```

Finally, we can just declare our command like this, and we are done! And again, you can just directly get the argument as a ready `IceCreamFlavor`
type without any additional parsing in the `executes(...)` method, which makes custom argument types very powerful.

```java
Commands.literal("icecream")
    .then(Commands.argument("flavor", new IceCreamArgument())
        .executes(ctx -> {
            final IceCreamFlavor flavor = ctx.getArgument("flavor", IceCreamFlavor.class);

            ctx.getSource().getSender().sendRichMessage("<b><red>Y<green>U<aqua>M<light_purple>!</b> You just had a scoop of <flavor>!",
                Placeholder.unparsed("flavor", flavor.toString())
            );
            return Command.SINGLE_SUCCESS;
        })
    )
    .build();
```

<div style={{display: 'inline-block', width: '100%'}}>
    <img src={IceCreamPng} style={{float: 'left', width: '100%'}}/>
</div>

> ## Note: Error handling during suggestions phase

Reviewer 2: Could be an admonition.
